packages/embark/src/cmd/cmd_controller.js

```diff
@@ -504,58 +504,73 @@ class EmbarkController {
+          await engine.events.request2("blockchain:node:start", engine.config.blockchainConfig, cb);
         });
-      },
-      function waitForWriteFinish(callback) {
-        // Skip if we are connected to a websocket, the server will do it
-        if(isSecondaryProcess(engine)) {
-          return callback();
-        }
-        engine.logger.info("Finished deploying".underline);
-        engine.events.once('outputDone', (err) => {
-          engine.logger.info(__("finished building").underline);
-          callback(err);
+        plugin.registerActionForEvent("embark:engine:started", async (_params, cb) => {
+          await engine.events.request2("storage:node:start", engine.config.storageConfig, cb);
+        });
+        plugin.registerActionForEvent("embark:engine:started", async (_params, cb) => {
+          await engine.events.request2("communication:node:start", engine.config.communicationConfig, cb);
+        });
+
+        engine.events.on('outputDone', function () {
+          engine.logger.info((__("Looking for documentation? You can find it at") + " ").cyan + "http://embark.status.im/docs/".green.underline + ".".cyan);
+          engine.logger.info(__("Ready").underline);
+          engine.events.emit("status", __("Ready").green);
+        });
+
+        engine.events.on('file-event', async ({ fileType, path }) => {
+          if (fileType === 'contract' || fileType === 'config') {
+            let contractsFiles = await engine.events.request2("config:contractsFiles");
+            let compiledContracts = await engine.events.request2("compiler:contracts:compile", contractsFiles);
+            let _contractsConfig = await engine.events.request2("config:contractsConfig");
+            let contractsConfig = cloneDeep(_contractsConfig);
+            let [contractsList, contractDependencies] = await engine.events.request2("contracts:build", contractsConfig, compiledContracts);
+            await engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies);
+          } else if (fileType === 'asset') {
+            engine.events.request('pipeline:generateAll', () => engine.events.emit('outputDone'));
+          }
+        });
+
+        engine.startEngine(async () => {
+          let contractsFiles = await engine.events.request2("config:contractsFiles");
+          let compiledContracts = await engine.events.request2("compiler:contracts:compile", contractsFiles);
+          let _contractsConfig = await engine.events.request2("config:contractsConfig");
+          let contractsConfig = cloneDeep(_contractsConfig);
+          let [contractsList, contractDependencies] = await engine.events.request2("contracts:build", contractsConfig, compiledContracts);
+          await engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies);
+          engine.events.request2("watcher:start")
+          callback();
```

Comment:
At this point it might makes sense to think about putting a lot of this stuff from this waterfall function into some other function that can be reused, since we're using a lot of the same logic in `run2()`, `build()` and now `console()`

Reply:
Totally agree, was gonna ask the same